features/mbedtls/targets/TARGET_NXP/aes_alt.c

```diff
@@ -0,0 +1,609 @@
+    MMCAU_AES_DecryptEcb(input, key, ctx->nr, output);
+#elif defined(MBEDTLS_FREESCALE_CAU3_AES)
+    if (!crypto_key_is_loaded(ctx))
+    {
+        CAU3_AES_SetKey(CAU3, &s_cau3Handle, key, ctx->nr);
+        crypto_attach_ctx_to_key_slot(ctx, s_cau3Handle.keySlot);
+    }
+    CAU3_AES_Decrypt(CAU3, &s_cau3Handle, input, output);
+#elif defined(MBEDTLS_FREESCALE_LPC_AES)
+    AES_SetKey(AES_INSTANCE, key, ctx->nr);
+    AES_DecryptEcb(AES_INSTANCE, input, output, 16);
+#elif defined(MBEDTLS_FREESCALE_CAAM_AES)
+    CAAM_AES_DecryptEcb(CAAM_INSTANCE, &s_caamHandle, input, output, 16, key, ctx->nr);
+#elif defined(MBEDTLS_FREESCALE_DCP_AES)
+    if (!crypto_key_is_loaded(ctx))
+    {
+        DCP_AES_SetKey(DCP, &s_dcpHandle, key, ctx->nr);
+        crypto_attach_ctx_to_key_slot(ctx, s_dcpHandle.keySlot);
+    }
+    DCP_AES_DecryptEcb(DCP, &s_dcpHandle, input, output, 16);
+#endif
+
+    return (0);
+}
+
+#if defined(MBEDTLS_CIPHER_MODE_CBC)
+/*
+ * AES-CBC buffer encryption/decryption
+ */
+#if defined(MBEDTLS_FREESCALE_LTC_AES)
+int mbedtls_aes_crypt_cbc(mbedtls_aes_context *ctx,
+                          int mode,
+                          size_t length,
+                          unsigned char iv[16],
+                          const unsigned char *input,
+                          unsigned char *output)
+{
+    uint8_t *key = (uint8_t *)ctx->rk;
+    uint32_t keySize = ctx->nr;
+
+    if (length % 16)
+        return (MBEDTLS_ERR_AES_INVALID_INPUT_LENGTH);
+
+    if (mode == MBEDTLS_AES_DECRYPT)
+    {
+        uint8_t tmp[16];
+        memcpy(tmp, input + length - 16, 16);
+        LTC_AES_DecryptCbc(LTC_INSTANCE, input, output, length, iv, key, keySize, kLTC_EncryptKey);
+        memcpy(iv, tmp, 16);
+    }
+    else
+    {
+        LTC_AES_EncryptCbc(LTC_INSTANCE, input, output, length, iv, key, keySize);
+        memcpy(iv, output + length - 16, 16);
+    }
+
+    return (0);
+}
+#elif defined(MBEDTLS_FREESCALE_DCP_AES)
+int mbedtls_aes_crypt_cbc(mbedtls_aes_context *ctx,
+                          int mode,
+                          size_t length,
+                          unsigned char iv[16],
+                          const unsigned char *input,
+                          unsigned char *output)
+{
+    uint8_t *key;
+
+    if (length % 16)
+        return (MBEDTLS_ERR_AES_INVALID_INPUT_LENGTH);
+
+    key = (uint8_t *)ctx->rk;
+    if (!crypto_key_is_loaded(ctx))
+    {
+        DCP_AES_SetKey(DCP, &s_dcpHandle, key, ctx->nr);
+        crypto_attach_ctx_to_key_slot(ctx, s_dcpHandle.keySlot);
+    }
+
+    if (mode == MBEDTLS_AES_DECRYPT)
+    {
+        uint8_t tmp[16];
+        memcpy(tmp, input + length - 16, 16);
+        DCP_AES_DecryptCbc(DCP, &s_dcpHandle, input, output, length, iv);
+        memcpy(iv, tmp, 16);
+    }
+    else
+    {
+        DCP_AES_EncryptCbc(DCP, &s_dcpHandle, input, output, length, iv);
+        memcpy(iv, output + length - 16, 16);
+    }
+
+    return (0);
+}
+#endif
```

Comment:
What about CBC when any of the following are set? How does CBC work on the following hardware?
 - MBEDTLS_FREESCALE_MMCAU_AES
 - MBEDTLS_FREESCALE_CAU3_AES
 - MBEDTLS_FREESCALE_LPC_AES
 - MBEDTLS_FREESCALE_CAAM_AES

Reply:
I have deleted MMCAU, CAU3, LPC and CAAM as this is not supported by this PR.